avoid re-applying views for every output when there is no pending mutations.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142118


In @bdhirsh words See https://github.com/pytorch/pytorch/issues/140835

---
It's due to this code:
```
storage = torch.empty(filesize,dtype=torch.uint8,)
torch.cat(swaps, out=storage)
swaps = storage.split(lengths)
return .... # return views of each tensor in swaps
```
When functionalized the traced program, we sync all graph outputs in case they have any pending mutations (code: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/functional_utils.py#L77)

That sync logic effectively says: if there is a pending mutation on one of the aliases, we need to regenerate the current (alias) tensor by replaying all of the views from the updated base tensor (code: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalTensorWrapper.cpp#L363)

In the case where one of those views is a split_with_sizes call, we will just... go ahead and run the view. We will do this for every graph output (which in this case, is all N views, each of which results in a split_with_size call.

In principle this is causing us to generate a graph with some unnecessary ops, that we could rely on the compiler to optimize. Although here, it sounds like all of these extra nodes are making a pretty big impact in the E2E compile time.

For this particular case, there is probably something simpler that we can do: there really shouldn't be any pending mutations on any of the graph outputs at the time that we return them. This is because when we first generated swaps in the example above, the mutation was already applied. There are not pending mutations at the time that we generate that view.
We actually had an old optimization that did it (originally for XLA) I have a PR that adds it back here 

---

laithsakka committed:
avoid replicating views

[ghstack-poisoned]

laithsakka commented on test/functorch/test_aotdispatch.py:
```diff
@@ -4924,7 +4919,7 @@ def forward(self, arg0_1):
             gm.true_graph_0.code.strip(),
             """\
-def forward(self, arg0_1):
+def forward(self, arg0_1, arg1_1, arg2_1):
```
we have more args here

laithsakka replied:
sounds not an issue those are passed in the call site above but then not used in the function 
cc @bdhirsh

bdhirsh replied:
hmm weirdly I tried running your PR locally and running this test, and I don't see this change. It's not immediately clear to me why the functionalization changes would affect what arguments cond() lifts during tracing

laithsakka replied:
strange?

laithsakka replied:
did build c++ again?

laithsakka commented on test/functorch/test_aotdispatch.py:
```diff
@@ -4914,7 +4909,7 @@ def forward(self, arg0_1):
     true_graph_0 = self.true_graph_0
     false_graph_0 = self.false_graph_0
-    cond = torch.ops.higher_order.cond(gt, true_graph_0, false_graph_0, [arg0_1]);  gt = true_graph_0 = false_graph_0 = arg0_1 = None
+    cond = torch.ops.higher_order.cond(gt, true_graph_0, false_graph_0, [arg0_1, 3, 4]);  gt = true_graph_0 = false_graph_0 = arg0_1 = None
```
we have 3, and 4 here as args we did not have before?

ezyang commented:
draft, so don't review yet?

laithsakka commented:
> draft, so don't review yet?

 I made it draft because I have to fix all the failures. they are still coming

laithsakka committed:
Update on "avoid re-applying views for every output when there is no pending mutations. "



In bdhirsh words See https://github.com/pytorch/pytorch/issues/140835

---
It's due to this code:
```
storage = torch.empty(filesize,dtype=torch.uint8,)
torch.cat(swaps, out=storage)
swaps = storage.split(lengths)
return .... # return views of each tensor in swaps
```
When functionalized the traced program, we sync all graph outputs in case they have any pending mutations (code: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/functional_utils.py#L77)

That sync logic effectively says: if there is a pending mutation on one of the aliases, we need to regenerate the current (alias) tensor by replaying all of the views from the updated base tensor (code: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalTensorWrapper.cpp#L363)

In the case where one of those views is a split_with_sizes call, we will just... go ahead and run the view. We will do this for every graph output (which in this case, is all N views, each of which results in a split_with_size call.

In principle this is causing us to generate a graph with some unnecessary ops, that we could rely on the compiler to optimize. Although here, it sounds like all of these extra nodes are making a pretty big impact in the E2E compile time.

For this particular case, there is probably something simpler that we can do: there really shouldn't be any pending mutations on any of the graph outputs at the time that we return them. This is because when we first generated swaps in the example above, the mutation was already applied. There are not pending mutations at the time that we generate that view.
We actually had an old optimization that did it (originally for XLA) I have a PR that adds it back here 

---

[ghstack-poisoned]

laithsakka committed:
Update on "avoid re-applying views for every output when there is no pending mutations. "



In bdhirsh words See https://github.com/pytorch/pytorch/issues/140835

---
It's due to this code:
```
storage = torch.empty(filesize,dtype=torch.uint8,)
torch.cat(swaps, out=storage)
swaps = storage.split(lengths)
return .... # return views of each tensor in swaps
```
When functionalized the traced program, we sync all graph outputs in case they have any pending mutations (code: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/functional_utils.py#L77)

That sync logic effectively says: if there is a pending mutation on one of the aliases, we need to regenerate the current (alias) tensor by replaying all of the views from the updated base tensor (code: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalTensorWrapper.cpp#L363)

In the case where one of those views is a split_with_sizes call, we will just... go ahead and run the view. We will do this for every graph output (which in this case, is all N views, each of which results in a split_with_size call.

In principle this is causing us to generate a graph with some unnecessary ops, that we could rely on the compiler to optimize. Although here, it sounds like all of these extra nodes are making a pretty big impact in the E2E compile time.

For this particular case, there is probably something simpler that we can do: there really shouldn't be any pending mutations on any of the graph outputs at the time that we return them. This is because when we first generated swaps in the example above, the mutation was already applied. There are not pending mutations at the time that we generate that view.
We actually had an old optimization that did it (originally for XLA) I have a PR that adds it back here 

---

[ghstack-poisoned]

laithsakka commented:
XLA change
https://github.com/pytorch/xla/pull/8458

laithsakka committed:
Update on "avoid re-applying views for every output when there is no pending mutations. "



In bdhirsh words See https://github.com/pytorch/pytorch/issues/140835

---
It's due to this code:
```
storage = torch.empty(filesize,dtype=torch.uint8,)
torch.cat(swaps, out=storage)
swaps = storage.split(lengths)
return .... # return views of each tensor in swaps
```
When functionalized the traced program, we sync all graph outputs in case they have any pending mutations (code: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/functional_utils.py#L77)

That sync logic effectively says: if there is a pending mutation on one of the aliases, we need to regenerate the current (alias) tensor by replaying all of the views from the updated base tensor (code: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalTensorWrapper.cpp#L363)

In the case where one of those views is a split_with_sizes call, we will just... go ahead and run the view. We will do this for every graph output (which in this case, is all N views, each of which results in a split_with_size call.

In principle this is causing us to generate a graph with some unnecessary ops, that we could rely on the compiler to optimize. Although here, it sounds like all of these extra nodes are making a pretty big impact in the E2E compile time.

For this particular case, there is probably something simpler that we can do: there really shouldn't be any pending mutations on any of the graph outputs at the time that we return them. This is because when we first generated swaps in the example above, the mutation was already applied. There are not pending mutations at the time that we generate that view.
We actually had an old optimization that did it (originally for XLA) I have a PR that adds it back here 

---

[ghstack-poisoned]

laithsakka commented:
Had a chat with  @bdhirsh he will take a look at test_advanced_indexing_correct_strides in two weeks or so.
regarding test_mutation_overlapping_mem we have another way to address that check but we need to make sure the issue in test_advanced_indexing_correct_strides is not major first.

bdhirsh commented:
Dug into it a little more, and it looks like we need this optimization in order to get correct strides. Example:
```
import torch

@torch.compile
def f(x):
    y = x.clone()[:, 1]
    z = torch.ones_like(y, dtype=torch.bool)
    y.masked_fill_(z, 0)
    return y

x = torch.randn(4, 4, device='cuda')
out = f(x)
print(out.stride())
```

With compile turned off, the stride we print is `(4,)`, but with compile on, and with the optimization patched in, we get `(1,)`.

You can see the difference in the ATen FX graphs:


Without the optimization (good)
```
 ===== Forward graph 0 =====
 /home/hirsheybar/local/b/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[4, 4][4, 1]cuda:0"):
         # File: /home/hirsheybar/local/b/pytorch/tmp3.py:5 in f, code: y = x.clone()[:, 1]
        clone: "f32[4, 4][4, 1]cuda:0" = torch.ops.aten.clone.default(arg0_1);  arg0_1 = None
        slice_1: "f32[4, 4][4, 1]cuda:0" = torch.ops.aten.slice.Tensor(clone, 0, 0, 9223372036854775807)
        select: "f32[4][4]cuda:0" = torch.ops.aten.select.int(slice_1, 1, 1);  slice_1 = None

         # File: /home/hirsheybar/local/b/pytorch/tmp3.py:6 in f, code: z = torch.ones_like(y, dtype=torch.bool)
        full: "b8[4][1]cuda:0" = torch.ops.aten.full.default([4], True, dtype = torch.bool, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)

         # File: /home/hirsheybar/local/b/pytorch/tmp3.py:7 in f, code: y.masked_fill_(z, 0)
        scalar_tensor: "f32[][]cuda:0" = torch.ops.aten.scalar_tensor.default(0.0, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0))
        where: "f32[4][1]cuda:0" = torch.ops.aten.where.self(full, scalar_tensor, select);  full = scalar_tensor = select = None
        slice_2: "f32[4, 4][4, 1]cuda:0" = torch.ops.aten.slice.Tensor(clone, 0, 0, 9223372036854775807)
        select_scatter: "f32[4, 4][4, 1]cuda:0" = torch.ops.aten.select_scatter.default(slice_2, where, 1, 1);  slice_2 = where = None
        slice_scatter: "f32[4, 4][4, 1]cuda:0" = torch.ops.aten.slice_scatter.default(clone, select_scatter, 0, 0, 9223372036854775807);  clone = select_scatter = None
        slice_3: "f32[4, 4][4, 1]cuda:0" = torch.ops.aten.slice.Tensor(slice_scatter, 0, 0, 9223372036854775807);  slice_scatter = None
        select_1: "f32[4][4]cuda:0" = torch.ops.aten.select.int(slice_3, 1, 1);  slice_3 = None
        return (select_1,)
```

with the optimization (bad):
```
 ===== Forward graph 0 =====
 /home/hirsheybar/local/a/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[4, 4][4, 1]cpu"):
         # File: /home/hirsheybar/local/a/pytorch/tmp.py:5 in f, code: y = x.clone()[:, 1]
        clone: "f32[4, 4][4, 1]cpu" = torch.ops.aten.clone.default(arg0_1);  arg0_1 = None
        slice_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.slice.Tensor(clone, 0, 0, 9223372036854775807);  clone = None
        select: "f32[4][4]cpu" = torch.ops.aten.select.int(slice_1, 1, 1);  slice_1 = None

         # File: /home/hirsheybar/local/a/pytorch/tmp.py:6 in f, code: z = torch.ones_like(y, dtype=torch.bool)
        full: "b8[4][1]cpu" = torch.ops.aten.full.default([4], True, dtype = torch.bool, layout = torch.strided, device = device(type='cpu'), pin_memory = False)

         # File: /home/hirsheybar/local/a/pytorch/tmp.py:7 in f, code: y.masked_fill_(z, 0)
        scalar_tensor: "f32[][]cpu" = torch.ops.aten.scalar_tensor.default(0.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
        where: "f32[4][1]cpu" = torch.ops.aten.where.self(full, scalar_tensor, select);  full = scalar_tensor = select = None
        return (where,)
```

The main issue is that if we have something like `return non_contiguous_tensor.inplace_op_(...)`, then when functionalization turns it into `return non_contiguous_tensor.out_of_place_op(...)`, we will allocate a fresh contiguous tensor for the output, which will have different strides than the eager code. Functionalization handles this today by doing something like:
```
base_tensor = ...
slice_tensor = slice_view_op(base_tensor, ...)
# functionalize the mutation
slice_tensor_updated = slice_tensor.out_of_place_op(...)
base_tensor_updated = slice_inverse(slice_tensor_updated, base_tensor, ...)
slice_tensor_updated_with_proper_strides = slice_view_op(base_tensor_updated, ...)
return slice_tensor_updated_with_proper_strides
```

and then relies on inductor reinplacing to make this efficient.

So unfortunately, I don't think we can just add this optimization back in without changing stride behavior :/

laithsakka commented:
closing this PR since this does not work.

bobrenjc93 commented:
After meeting notes: Check if contiguous and only do optimization if so

bobrenjc93 commented:
Alternate (harder to implement) solution: Whenever we see a sequence of outputs that come from the same multi-view op, then do clustering and only do the view replay once.
